websocket-chat-broker/src/main.rs

```diff
@@ -0,0 +1,156 @@
+#[macro_use]
+extern crate actix;
+extern crate actix_web;
+extern crate actix_broker;
+extern crate futures;
+extern crate rand;
+
+use actix::prelude::*;
+use actix::fut;
+use actix_web::server::HttpServer;
+use actix_web::{fs, http, ws, App, Error, HttpRequest, HttpResponse};
+use actix_broker::BrokerIssue;
+use futures::Future;
+
+mod server;
+use server::*;
+
+fn chat_route(req: &HttpRequest<()>) -> Result<HttpResponse, Error> {
+    ws::start(req, WsChatSession::default())
+}
+
+#[derive(Default)]
+struct WsChatSession {
+    id: usize,
+    room: String,
+    name: Option<String>,
+}
+
+impl WsChatSession {
+    fn join_room(&mut self, room_name: &str, ctx: &mut ws::WebsocketContext<Self>) {
+        let room_name = room_name.to_owned();
+        // First send a leave message for the current room
+        let leave_msg = LeaveRoom(self.room.clone(), self.id);
+        self.issue_sync(leave_msg, ctx);
+        // Then send a join message for the new room
+        let join_msg = JoinRoom(
+            room_name.to_owned(),
+            self.name.clone(),
+            ctx.address().recipient());
+
+        WsChatServer::from_registry()
+            .send(join_msg)
+            .map_err(|_| ())
+            .into_actor(self)
+            .map(|id, act, _| {
+                act.id = id;
+                act.room = room_name;
+            })
+            .wait(ctx);
```

Comment:
It would be nice to avoid blocking in async code.
It is it possible to avoid relying on wait?

P.s. it is general comment for all futures that are transformed to ActorFuture

Reply:
Thanks for the hint. You were dead right - both those futures could be spawned instead of waited upon.